sql/sort.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// orderBy constructs a sortNode based on the ORDER BY clause. Construction of
+// the sortNode might adjust the number of render targets in the scanNode if
+// any ordering expressions are specified.
+func (p *planner) orderBy(n *parser.Select, s *scanNode) (*sortNode, error) {
+	if n.OrderBy == nil {
+		return nil, nil
+	}
+
+	// We grab a copy of columns here because we might add new render targets
+	// below. This is the set of columns requested by the query.
+	columns := s.Columns()
+	var ordering []int
+
+	for _, o := range n.OrderBy {
+		index := 0
+		switch t := o.Expr.(type) {
+		case *parser.QualifiedName:
+			if len(t.Indirect) > 0 {
+				return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+			}
+			target := string(t.Base)
+			for j, col := range columns {
+				if equalName(target, col) {
+					index = j + 1
+					break
+				}
+			}
+			if index == 0 {
+				return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+			}
+
+		default:
+			// TODO(pmattis): Use NormalizeExpr.
+			datum, err := parser.EvalExpr(o.Expr)
+			if err != nil {
```

Comment:
Yeah, I'm assuming that `EvalExpr` returned an error because the expression contained a qualified name. Once `NormalizeExpr` is used we can return the error instead. I've added a comment to this effect.